### PR TITLE
EL-1909 Add a single category search page

### PR DESCRIPTION
**Pull request by willc-work**
## What does this pull request do?

add a feature flag to hide this new feature in STG and PROD and show it in UAT
create new search view with dynamic content and URL depending on category passed in the URL
handle results and display on results page
add dynamic results page content

Does not:
add a back button to results page
does not do region check for error handling

further tickets will be created to undertake the other work.

https://dsdmoj.atlassian.net/browse/EL-1967
https://dsdmoj.atlassian.net/browse/EL-1980
https://dsdmoj.atlassian.net/browse/EL-1981
https://dsdmoj.atlassian.net/browse/EL-1982
https://dsdmoj.atlassian.net/browse/EL-1983
https://dsdmoj.atlassian.net/browse/EL-1984

## Any other changes that would benefit highlighting?

Intentionally left blank.

## Checklist

- [ ] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: fala/templates/adviser/single_category_search.html [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+{% import "macros/element.html" as Element %}
+{% import "macros/forms.html" as Form %}
+
+{%- from 'govuk_frontend_jinja/components/error-summary/macro.html' import govukErrorSummary %}
+{%- from 'govuk_frontend_jinja/components/checkboxes/macro.html' import govukCheckboxes %}
+{%- from "govuk_frontend_jinja/components/input/macro.html" import govukInput %}
+
+{% block content %}
+  {% if results %}
+    <!-- Content for search results -->
```
Comment (MazOneTwoOne) [inline fala/templates/adviser/single_category_search.html:20]: Question - why aren't we using the `results.html`? Is that part of another ticket

File: fala/apps/adviser/views.py [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 from django.http import HttpResponse
 import os
 from django.views import View
-
 from .forms import AdviserSearchForm, AdviserRootForm
-from .laa_laa_paginator import LaaLaaPaginator
 from .regions import Region
+from django.shortcuts import redirect, render
+from .models import EnglandOrWalesState, ErrorState, OtherJurisdictionState
+from .forms import SingleCategorySearchForm
```
Comment (MazOneTwoOne) [inline fala/apps/adviser/views.py:12]: You can move this to line 8

File: fala/apps/adviser/forms.py [hunk 3]
```diff
@@ -193,3 +192,141 @@ def search(self):
...
+    )
+
+    def __init__(self, categories=None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Ensure categories is a list and assign it to self.categories
+        self.categories = categories if categories is not None else []
+        self._region = None
+        self._country_from_valid_postcode = None
+
+    def clean_postcode(self):
```
Comment (MazOneTwoOne) [inline fala/apps/adviser/forms.py:212]: could be wrong but I don't think you need `clean_postcode` the `clean` method already does this?

File: fala/apps/adviser/utils.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# utils.py
+
+CATEGORY_TRANSLATIONS = {
```
Comment (psweeting1) [inline fala/apps/adviser/utils.py:3]: This CATEGORY_TRANSLATIONS can be moved out into it owns python file. 

Let's make a new directory called constants under apps, so `fala/apps/constants`

Here we'll make a python file call `categories.py`

File: fala/apps/adviser/utils.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    "hlpas": "Tell the adviser your home is at risk and you want advice through the Housing Loss Prevention Advice Service.",
+    "welfare-benefits": (
+        "Legal aid for advice about welfare benefits is only available if you are appealing a decision made by the social security tribunal. "
+        "This appeal must be in the Upper Tribunal, Court of Appeal or Supreme Court.\n\n"
+        "For any other benefits issue, ask the legal adviser if you can get free legal advice or if you will have to pay for it."
+    ),
+    "clinical-negligence": "Legal aid for advice about clinical negligence is usually only available if you have a child that suffered a brain injury during pregnancy, birth or the first 8 weeks of life.",
+}
+
+CATEGORY_DISPLAY_NAMES = {
```
Comment (psweeting1) [inline fala/apps/adviser/utils.py:44]: So you can add this to the Category translation you have like this:

`"hlpas": ["hlpas", "Housing Loss Prevention Advice Service"]`

```
values = CATEGORY_TRANSLATIONS["hlpas"]
print(values[0])  # Output: hlpas
print(values[1])  # Output: Housing Loss Prevention Advice Service
```

You could then also add the full name of each of the categories so you don't have to format them into something readable. 

What do you think?

File: fala/apps/adviser/utils.py (hunk 4 above)
Comment (psweeting1) [reply fala/apps/adviser/utils.py:3]: Don't forget to place an `__init__.py` into the directory. It helps with imports with Python.

File: fala/templates/adviser/single_category_search.html (hunk 1 above)
Comment (willc-work) [reply fala/templates/adviser/single_category_search.html:20]: i just wanted to get some results on display, we can write a ticket to make sure results template is used by this search I was just trying to go with the most simplified results page I could to display the data I was getting bacl

File: fala/apps/adviser/utils.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    "hlpas": "hlpas",
+    "immas": "immigration-asylum",
+    "mhe": "mental-health",
+    "mosl": "modern-slavery",
+    "pl": "prison-law",
+    "pub": "public-law",
+    "wb": "welfare-benefits",
+}
+
+SLUG_TO_CODE = {v: k for k, v in CATEGORY_TRANSLATIONS.items()}
```
Comment (psweeting1) [inline fala/apps/adviser/utils.py:23]: This is me, but I dislike single character variables, they aren't descriptive and sometime make it harder to read. Using descriptive variable names in loops enhances readability and makes the code more maintainable.

```
SLUG_TO_CODE = {display_name: category_code for category_code, display_name in CATEGORY_TRANSLATIONS.items()}
```

File: fala/apps/adviser/utils.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+def get_category_code_from_slug(slug):
+    return SLUG_TO_CODE.get(slug)
+
+
+def get_category_display_name(category_code):
+    return CATEGORY_TRANSLATIONS.get(category_code)
+
+
+CATEGORY_MESSAGES = {
```
Comment (psweeting1) [inline fala/apps/adviser/utils.py:34]: Again on this this python file into it's own. Something like `constants/category_messages.py`

This doesn't need to be a class, just something that can be isolated into the app that's using it.

File: fala/apps/adviser/models.py [hunk 8]
```diff
@@ -10,3 +14,145 @@ class SatisfactionFeedback(models.Model):
 
     def __str__(self):
         return f"Feedback {self.id}"
+
+
+class EnglandOrWalesState(object):
```
Comment (psweeting1) [inline fala/apps/adviser/models.py:19]: With classes that we want to re-use we can do the following:

```
fala/
├── common/
│   ├── __init__.py
│   ├── states.py  # New file for reusable classes like EnglandOrWalesState
├── app/adviser/
│   ├── views.py
├── app2/single-page/
│   ├── views.py
```

common will be a directory that we will use to access anything we need across all of our apps.

File: fala/apps/adviser/forms.py [hunk 9]
```diff
@@ -193,3 +192,141 @@ def search(self):
...
+        # Check if categories are provided
+        if not categories:
+            self.add_error("categories", _("Category is required."))
+        else:
+            self.categories = [categories]
+
+        return cleaned_data
+
+    @property
+    def region(self):
```
Comment (psweeting1) [inline fala/apps/adviser/forms.py]: Out of scope of this PR I know, but just a note.

We can refactor this to be in it's own python file and be a class under constants directory (More on directory below in the PR).

This is going to a big theme of these tickets to move them out into single methods in their own python constant file, help functions (Utils) or if deem something with scalability in mind, a class. 

For anything small, like a helper function, validator, formatter `utils.py` is where you would move those type of functions so you can import and re-use them. If there is a case where you want to use it anywhere in the app. The common directory is the best place to create a new `utils.py` python file to store them. (Again see further down in the PR where I talk about common directory)

File: fala/apps/adviser/models.py (hunk 8 above)
Comment (willc-work) [reply fala/apps/adviser/models.py:19]: created a ticket for this refactor:
https://dsdmoj.atlassian.net/browse/EL-1981

File: fala/apps/adviser/utils.py (hunk 5 above)
Comment (willc-work) [reply fala/apps/adviser/utils.py:58]: created a ticket for this refactor:
https://dsdmoj.atlassian.net/browse/EL-1980

**Review by psweeting1 — APPROVED** [review]
LGTM

File: fala/apps/adviser/views.py [hunk 10]
```diff
@@ -248,6 +168,7 @@ def get_queryset(self):
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
+        context["is_single_category_search"] = self.request.GET.get("is_single_category_search", False)
```
Comment (psweeting1) [inline fala/apps/adviser/views.py:171]: Does this not make the URL `/search?is_single_category_search=True?`

Comment (MazOneTwoOne) [reply fala/apps/adviser/views.py]: Yes, it'll be `/search?categories=hlpas&single_category_search_redirect=true&postcode=se1`

**Pull request merged**